README.md
# NAME

Slack::RTM::Bot - This is a perl module helping to create slack bot with Real Time Messaging(RTM) API.

# SYNOPSIS

    use Slack::RTM::Bot;
    my $bot = Slack::RTM::Bot->new( token => '<API token>');

    $bot->on({
            channel => 'general',
            text    => qr/.*/
        },
        sub {
            my ($response) = @_;
            print $response->{text}."\n";
        }
    );

    $bot->start_RTM(sub {

        $bot->say(
            channel => 'general',
            text    => '<!here> hello, world.'
        );

        $bot->say(
            channel => '@username',
            text    => 'hello, world.'
        );

        while(1) { sleep 10; print "I'm not dead\n"; }
    });

# METHODS

## new

    method new(token => $token)

Constructs a [Slack::RTM::Bot](https://metacpan.org/pod/Slack%3A%3ARTM%3A%3ABot) object.

The `$token` is the slack API token.

## on

    method on(\%event, $callback)

`$callback` will be executed when it fitted the `\%event` conditions.
The `\%event` key is equal to a key of json received from slack, and value is estimated as regex.

`$callback` is handed JSON object of message received from Slack.

## start\_RTM

    method start_RTM($callback)

It start Real Time Messaging API.
`$callback` will be executed when establish connection.
`start_RTM` make child process. Thus, you must call `stop_RTM` if you want to kill child processes before stop main process.

## stop\_RTM

    method stop_RTM()

It stop Real Time Messaging API.

## say

    method say(%options)

It sent a message to a Slack. The channel name can be used to designate channel.
if you want to send a direct message, let designate the @username as a channel.

# SOURCE CODE

This is opensource software.

https://github.com/duck8823/Slack-RTM-Bot

# HOW TO CONTRIBUTE

## with installing
The fastest way to get started working with the code is to run the following commands:

    $ git clone https://github.com/duck8823/Slack-RTM-Bot.git
    $ cd Slack-RTM-Bot
    $ cpanm --installdeps .
    $ perl Build.PL
    $ ./Build
    $ ./Build install
    $ ./Build test  # run the tests

## without installing
or without installing Slack-RTM-Bot, run the following commands:

    $ git clone https://github.com/duck8823/Slack-RTM-Bot.git
    $ cd Slack-RTM-Bot
    $ cpanm --installdeps .  # install dependencies

and run your script with \`-I/path/to/Slack-RTM-Bot/lib\` option.

    $ perl -I/path/to/Slack-RTM-Bot/lib your_script.pl

# SEE ALSO

https://api.slack.com/rtm

# LICENSE

The MIT License (MIT)

Copyright (c) 2016 Shunsuke Maeda

See LICENSE file.

# AUTHOR

Shunsuke Maeda <duck8823@gmail.com>


Changes
Revision history for Perl extension Slack-RTM-Bot

{{$NEXT}}

1.15 2022-11-13T06:40:31Z

  - #38 Users -> Members fix
  - #40 Use rtm connect api
  - #41 fix failed to send direct messages
  - #42 Use rtm.connect instead of rtm.start

1.13 2020-07-05T15:03:42Z

  - #35 Use conversations.list

1.12 2019-05-04T01:44:52Z

   - #30 Avoid getting killed by a bug in Slack.

1.11 2019-04-27T00:39:51Z

   - #29 Slack changed the limit maximum 
   - #28 Added missing license meta

1.10 2019-04-17T11:56:22Z

   - Fix parameter key

1.09 2019-04-13T10:17:17Z

   - Clarify a few die() errors
   - Fix missing conversations

1.08 2019-01-20T00:39:26Z

   - Migrate channels/groups to conversations api

1.07 2019-01-10T22:57:02Z

   - Fix url getting channels

1.06 2018-12-29T14:19:45Z

   - Fix a bug that spawn zombies without stop_RTM

1.05 2017-05-20T05:01:16Z

   - Fix channel or group not found

1.04 2017-04-08T03:30:26Z

   - Fix dying when response to new channel/group

1.03 2017-04-05T23:11:07Z

   - Fix a bug that zombies spawn when stop_RTM
   - Fix a problem that main process stop when die in callback process

1.02 2017-03-29T23:12:01Z

   - Add x_static_install flag

1.01 2017-03-25T00:20:55Z

   - Improve that max_message_size can be set to Protocol::WebSocket::Client
   - Remove reconnect (not working method)

1.00 2017-03-15T17:07:06Z

   - Restore threads to fork (but on MSWin32, using threads)
   - Fix to wait establish connection on start_RTM

0.14 2017-02-18T16:06:25Z

   - Fix tests without threads

0.13 2017-02-11T04:35:24Z

    - Using thread instead of fork
    - Fixes say()

0.12 2016-10-08T14:29:19Z

    - Fixes reading/writing to private channels 

0.11 2016-10-05T14:23:02Z

    - Add a error message
    - Add a How To Contribute section in POD

0.10 2016-09-17T08:15:46Z

    - Fix synopsis

0.09 2016-08-28T06:31:34Z

    - Add options

0.08 2016-08-04T18:04:39Z

    - Fix action matching process

0.07 2016-06-25T00:26:06Z

    - Fix Garbled characters
    - Improve execution timing of "say" method

0.06 2016-06-24T00:55:32Z

    - Fix a bug that isn't working without using "say" 

0.05 2016-06-23T16:13:50Z

    - Fix a bug that isn't working on Linux

0.04 2016-06-23T12:09:44Z

    - Fix cpanfile

0.03 2016-06-22T23:04:39Z

    - Turned off verifying the hostname.

0.02 2016-06-22T15:55:41Z

    - Fix a bug that child process will be zombie without stop_RTM.

0.01 2016-06-21T22:20:24Z

    - original version



lib/Slack/RTM/Bot/Client.pm
package Slack::RTM::Bot::Client;

use strict;
use warnings;

use JSON;
use Encode;
use Data::Dumper;
use HTTP::Request::Common qw(POST GET);
use LWP::UserAgent;
use LWP::Protocol::https;

use Protocol::WebSocket::Client;
use IO::Socket::SSL qw/SSL_VERIFY_NONE/;

use Slack::RTM::Bot::Information;
use Slack::RTM::Bot::Response;

my $ua = LWP::UserAgent->new(
	ssl_opts => {
		verify_hostname => 0,
		SSL_verify_mode => SSL_VERIFY_NONE
	}
);
$ua->agent('Slack::RTM::Bot');

sub new {
	my $pkg = shift;
	my $self = {
		@_
	};
	die "token is required." unless $self->{token};
	return bless $self, $pkg;
}

sub connect {
	my $self = shift;
	my ($token) = @_;

	my $res = $ua->request(POST 'https://slack.com/api/rtm.connect', [ token => $token ]);
	my $content;
	eval {
		$content = JSON::from_json($res->content);
	};
	if ($@) {
		die 'connect response fail:'.Dumper $res->content;
	}
	die 'connect response fail: '.$res->content unless ($content->{ok});

	$self->{info} = Slack::RTM::Bot::Information->new(%{$content});
	$self->_connect;
}

sub _connect {
	my $self = shift;
	my ($host) = $self->{info}->{url} =~ m{wss://(.+)/websocket};
	my $socket = IO::Socket::SSL->new(
		SSL_verify_mode => SSL_VERIFY_NONE,
		PeerHost => $host,
		PeerPort => 443
	);
	$socket->blocking(0);
	$socket->connect;

	my $ws_client = Protocol::WebSocket::Client->new(url => $self->{info}->{url});
	$ws_client->{hs}->req->{max_message_size} = $self->{options}->{max_message_size} if $self->{options}->{max_message_size};
	$ws_client->{hs}->res->{max_message_size} = $self->{options}->{max_message_size} if $self->{options}->{max_message_size};
	$ws_client->on(read => sub {
			my ($cli, $buffer) = @_;
			$self->_listen($buffer);
		});
	$ws_client->on(write => sub {
			my ($cli, $buffer) = @_;
			syswrite $socket, $buffer;
		});
	$ws_client->on(connect => sub {
			print "RTM (re)connected.\n" if ($self->{options}->{debug});
		});
	$ws_client->on(error => sub {
			my ($cli, $error) = @_;
			print STDERR 'error: '. $error;
		});
	$ws_client->connect;

	$self->{ws_client} = $ws_client;
	$self->{socket} = $socket;
}

sub disconnect {
	my $self = shift;
	$self->{ws_client}->disconnect;
	undef $self;
}

sub read {
	my $self = shift;
	my $data = '';
	while (my $line = readline $self->{socket}) {
		$data .= $line;
	}
	if ($data) {
		$self->{ws_client}->read($data);
		return $data =~ /.*hello.*/;
	}
}

sub write {
	my $self = shift;
	$self->{ws_client}->write(JSON::encode_json({@_}));
}

sub find_conversation_id {
	my $self = shift;
	my ($name) = @_;
	my $id = $self->{info}->_find_conversation_id($name);
	$id ||= $self->_refetch_conversation_id($name) or die "There are no conversations of such name: $name";
	return $id;
}

sub _refetch_conversation_id {
	my $self = shift;
	my ($name) = @_;
	$self->_refetch_conversations;
	return $self->{info}->_find_conversation_id($name);
}

sub find_conversation_name {
	my $self = shift;
	my ($id) = @_;
	my $name = $self->{info}->_find_conversation_name($id);
	$name ||= $self->_refetch_conversation_name($id) or warn "There are no conversations of such id: $id";
	$name ||= $id;
	return $name;
}

sub _refetch_conversation_name {
	my $self = shift;
	my ($id) = @_;
	$self->_refetch_conversations;
	return $self->{info}->_find_conversation_name($id);
}

sub _refetch_conversations {
	my $self = shift;
	my $cursor = "";
	do {
		my $res = $ua->request(POST 'https://slack.com/api/conversations.list', [ token => $self->{token}, types => "public_channel,private_channel,im", cursor => $cursor ]);
		my $content;
		eval {
			$content = JSON::decode_json($res->content);
		};
		if ($@) {
			die 'connect response fail:' . Dumper $res->content;
		}
		die 'connect response fail: ' . $res->content unless ($content->{ok});

		for my $channel (@{$content->{channels}}) {
			if ($channel->{is_im}) {
				my $user_id = $channel->{user};
				my $name = $self->{info}->_find_user_name($user_id);
				$name ||= $self->_refetch_user_name($user_id) or warn "There are no users of such id: $user_id";
				$self->{info}->{channels}->{$channel->{id}} = { %$channel, name => '@'.$name };
				next;
			}
			$self->{info}->{channels}->{$channel->{id}} = $channel;
		}

		$cursor = $content->{response_metadata}->{next_cursor};
	} until ($cursor eq "");
}

sub find_user_name {
	my $self = shift;
	my ($id) = @_;
	my $name = $self->{info}->_find_user_name($id);
	$name ||= $self->_refetch_user_name($id) or warn "There are no users of such id: $id";
	$name ||= $id;
	return $name;
}

sub _refetch_user_id {
	my $self = shift;
	my ($name) = @_;
	$self->_refetch_users;
	return $self->{info}->_find_user_id($name);
}

sub _refetch_user_name {
	my $self = shift;
	my ($id) = @_;
	$self->_refetch_users;
	return $self->{info}->_find_user_name($id);
}

sub _refetch_users {
	my $self = shift;
	my $res;
	eval {
		my $users = {};
		my $cursor = "";
		do {
			$res = $ua->request(GET "https://slack.com/api/users.list?token=$self->{token}&cursor=$cursor");
			my $args = JSON::from_json($res->content);
			for my $user (@{$args->{members}}) {
				$users->{$user->{id}} = $user;
			}
			if (defined($args->{response_metadata}->{next_cursor})) {
				$cursor = $args->{response_metadata}->{next_cursor};
			}
		} until ($cursor eq "");
		$self->{info}->{users} = $users;
       };
       if ($@) {
	       die '_refetch_users response fail:'.Dumper $res->content;
       }
}

sub _listen {
	my $self = shift;
	my ($buffer) = @_;
	my $buffer_obj;
	eval {
		$buffer_obj = JSON::from_json($buffer);
	};
	if ($@) {
		die "response is not json string. : $buffer";
	}
	if ($buffer_obj->{type} && $buffer_obj->{type} eq 'reconnect_url') {
		$self->{info}->{url} = $buffer_obj->{url};
	}

	my ($user, $channel);
	if ($buffer_obj->{user} && !ref($buffer_obj->{user})) {
		$user = $self->find_user_name($buffer_obj->{user});
		warn "There are no users of such id: $buffer_obj->{user}" unless $user;
	}
	if ($buffer_obj->{channel} && !ref($buffer_obj->{channel})) {
		$channel = $self->find_conversation_name($buffer_obj->{channel});
		warn "There are no conversations of such id: $buffer_obj->{channel}" unless $channel;

	}
	my $response = Slack::RTM::Bot::Response->new(
		buffer  => $buffer_obj,
		user    => $user,
		channel => $channel
	);
ACTION: for my $action(@{$self->{actions}}){
		for my $key(keys %{$action->{events}}){
			my $regex = $action->{events}->{$key};
			if(!defined $response->{$key} || $response->{$key} !~ $regex){
				next ACTION;
			}
		}
		eval {
			$action->{routine}->($response);
		};
		if ($@) {
			warn $@;
			kill 9, @{$self->{pids}};
			exit(1);
		}
	}
};

1;


lib/Slack/RTM/Bot/Information.pm
package Slack::RTM::Bot::Information;

use strict;
use warnings;

sub new {
	my $pkg = shift;
	my ($args) = {@_};
	my $self = {
		@_,
	};
	$self->{users}    = &_parse_users($args);
	$self->{channels} = &_parse_conversations($args);
	return bless $self, $pkg;
}

sub _parse_users {
	my $args = shift;
	my $users = {};
	for my $user (@{$args->{users}}){
		$users->{$user->{id}} = $user;
	}
	return $users;
}

sub _parse_conversations {
	my $args = shift;
	my $conversations = {};
	for my $conversation (@{$args->{channels}}){
		$conversations->{$conversation->{id}} = $conversation;
	}
	return $conversations;
}

sub _find_conversation_id {
	my $self = shift;
	my ($name) = @_;
	my $conversations = $self->{channels};

	for my $key (keys %{$conversations}){
		if($name eq $conversations->{$key}->{name}){
			return $conversations->{$key}->{id};
		}
	}
	return undef;
}

sub _find_conversation_name {
	my $self = shift;
	my ($id) = @_;
	my $conversations = $self->{channels};
	return $conversations->{$id}->{name} if $conversations->{$id};
}

sub _find_user_name {
	my $self = shift;
	my ($id) = @_;
	my $users = $self->{users};
	return $users->{$id}->{name} if $users->{$id};
}

1;


lib/Slack/RTM/Bot.pm
package Slack::RTM::Bot;

use 5.008001;
use strict;
use warnings;

use POSIX qw/sys_wait_h/;

use JSON;
use Slack::RTM::Bot::Client;

our $VERSION = "1.15";

pipe(READH, WRITEH);
select(WRITEH);$|=1;
pipe(READH2, WRITEH2);
select(WRITEH2);$|=1;
select(STDOUT);

sub new {
	my $pkg = shift;
	my $self = {
		@_
	};
	die 'need token!' unless $self->{token};
	return bless $self, $pkg;
}

sub start_RTM {
	my $self = shift;
	my ($sub) = @_;
	$self->_connect($self->{options});

	my $parent = $$;

	if ($^O ne 'MSWin32') {
		my @children = ();

		my $pid = fork;
		push @children, $pid;
		unless ($pid) {
			while (1) {
				unless (kill 0, $pid) {
					kill 9, $pid;
					waitpid($pid, WUNTRACED);
					last;
				}
				print WRITEH "\n";
				sleep 1;
			}
		} else {
			my $pid = fork;
			push @children, $pid;
			unless ($pid) {
				$self->{client}->{pids} = [$parent, @children];
				my $i = 0;
				while (1) {
					unless (kill 0, $parent) {
						kill 9, $pid;
						waitpid($pid, WUNTRACED);
						last;
					}
					if ($self->{client}->read) {
						print WRITEH2 "\n";
					}
					(my $buffer = <READH>) =~ s/\n.*$//;
					if ($buffer) {
						$self->{client}->write(
							%{JSON::from_json(Encode::decode_utf8($buffer))}
						);
					}
					if (++$i % 30 == 0) {
						$self->{client}->write(
							id   => $i,
							type => 'ping'
						);
					}
				}
			} else {
				$self->{children} = \@children;
				# wait until connected
				<READH2>;
				&$sub($self) if $sub;
			}
		};
	} else {
		require threads;
		require Thread::Queue;

		threads->create(
			sub {
				while (kill 0, $parent) {
					print WRITEH "\n";
					sleep 1;
				}
			}
		)->detach;

		threads->create(
			sub {
				my $i = 0;
				while (kill 0, $parent) {
					if ($self->{client}->read) {
						print WRITEH2 "\n";
					}
					(my $buffer = <READH>) =~ s/\n.*$//;
					if ($buffer) {
						$self->{client}->write(
							%{JSON::from_json(Encode::decode_utf8($buffer))}
						);
					}
					if (++$i % 30 == 0) {
						$self->{client}->write(
							id   => $i,
							type => 'ping'
						);
					}
				}
			}
		)->detach;

		$self->{queue} = Thread::Queue->new();
		$self->{worker} = threads->create(sub {
			while (defined(my $req = $self->{queue}->dequeue())) {
				print WRITEH $req;
			}
		});

		# wait until connected
		<READH2>;
		&$sub($self) if $sub;
	}
}

sub stop_RTM {
	my $self = shift;

	sleep 1;
	$self->{client}->disconnect;
	undef $self->{client};

	if ($^O ne 'MSWin32') {
		for my $child (@{$self->{children}}) {
			kill 9, $child;
			waitpid($child, WUNTRACED);
		}
		undef $self->{children};
	} else {
		$self->{queue}->end();
		$self->{worker}->join();
	}
}

sub _connect {
	my $self = shift;

	my $client = Slack::RTM::Bot::Client->new(
		token   => $self->{token},
		actions => $self->{actions},
		options => $self->{options}
	);
	$client->connect($self->{token});

	$self->{client} = $client;
}

sub say {
	my $self = shift;
	my $args;
	if(!@_ || scalar @_ % 2 != 0) {
		die "argument is not a HASH or ARRAY."
	}
	$args = {@_};
	if(!defined $args->{text} || !defined $args->{channel}) {
		die "argument needs keys 'text' and 'channel'.";
	}

	die "RTM not started." unless $self->{client};

	my $request = JSON::to_json({
		type    => 'message',
		subtype => 'bot_message',
		bot_id  => $self->{client}->{info}->{self}->{id},
		%$args,
		channel => $self->{client}->find_conversation_id($args->{channel})
	})."\n";
	print WRITEH $request;
}

sub on {
	my $self = shift;
	die "RTM already started." if $self->{info};
	my ($events, $routine) = @_;
	push @{$self->{actions}}, {
			events  => $events,
			routine => $routine
		};
}

sub add_action {
	my $self = shift;
	$self->on(@_);
}

1;
__END__

=encoding utf-8

=head1 NAME

Slack::RTM::Bot - This is a perl module helping to create slack bot with Real Time Messaging(RTM) API.

=head1 SYNOPSIS

    use Slack::RTM::Bot;
    my $bot = Slack::RTM::Bot->new( token => '<API token>');

    $bot->on({
            channel => 'general',
            text    => qr/.*/
        },
        sub {
            my ($response) = @_;
            print $response->{text}."\n";
        }
    );

    $bot->start_RTM(sub {

        $bot->say(
            channel => 'general',
            text    => '<!here> hello, world.'
        );

        $bot->say(
            channel => '@username',
            text    => 'hello, world.'
        );

        while(1) { sleep 10; print "I'm not dead\n"; }
    });

=head1 METHODS

=head2 new

  method new(token => $token)

Constructs a L<Slack::RTM::Bot> object.

The C<$token> is the slack API token.

=head2 on

  method on(\%event, $callback)

C<$callback> will be executed when it fitted the C<\%event> conditions.
The C<\%event> key is equal to a key of json received from slack, and value is estimated as regex.

C<$callback> is handed JSON object of message received from Slack.

=head2 start_RTM

  method start_RTM($callback)

It start Real Time Messaging API.
C<$callback> will be executed when establish connection.
C<start_RTM> make child process. Thus, you must call C<stop_RTM> if you want to kill child processes before stop main process.

=head2 stop_RTM

  method stop_RTM()

It stop Real Time Messaging API.

=head2 say

  method say(%options)

It sent a message to a Slack. The channel name can be used to designate channel.
if you want to send a direct message, let designate the @username as a channel.

=head1 SOURCE CODE

This is opensource software.

https://github.com/duck8823/Slack-RTM-Bot

=head1 HOW TO CONTRIBUTE

=head2 with installing
The fastest way to get started working with the code is to run the following commands:

  $ git clone https://github.com/duck8823/Slack-RTM-Bot.git
  $ cd Slack-RTM-Bot
  $ cpanm --installdeps .
  $ perl Build.PL
  $ ./Build
  $ ./Build install
  $ ./Build test  # run the tests

=head2 without installing
or without installing Slack-RTM-Bot, run the following commands:

  $ git clone https://github.com/duck8823/Slack-RTM-Bot.git
  $ cd Slack-RTM-Bot
  $ cpanm --installdeps .  # install dependencies

and run your script with `-I/path/to/Slack-RTM-Bot/lib` option.

  $ perl -I/path/to/Slack-RTM-Bot/lib your_script.pl

=head1 SEE ALSO

https://api.slack.com/rtm

=head1 LICENSE

The MIT License (MIT)

Copyright (c) 2016 Shunsuke Maeda

See LICENSE file.

=head1 AUTHOR

Shunsuke Maeda E<lt>duck8823@gmail.comE<gt>

=cut

